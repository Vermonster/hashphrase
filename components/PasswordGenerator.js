import React from 'react';
import { Keyboard, Clipboard, View } from 'react-native';
import loplop from 'loplop';
import FormInput from './FormInput';
import BaseButton from './BaseButton';

export default class PasswordGenerator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      label: '',
      password: '',
    };
  }

    handleChange = name => (e) => {
      const inputValue = e.nativeEvent.text;
      this.setState({ [name]: inputValue });
    };

    addToClipboard = (password) => {
      const { onClipboardSave } = this.props;
      Clipboard.setString(password);
      onClipboardSave();
    };

    generatePassword = (e) => {
      e.preventDefault();
      const { label, password } = this.state;
      this.setState({ generatedPassword: loplop(label, password) }, () => {
        const { generatedPassword } = this.state;
        this.addToClipboard(generatedPassword);
      });
      Keyboard.dismiss();
    };

    render() {
      const { label, password } = this.state;
      return (
        <View>
          <FormInput name="label" value={label} placeholder="Label..." onChanged={this.handleChange('label')} />
          <FormInput name="password" value={password} placeholder="Master Password..." secureTextEntry onChanged={this.handleChange('password')} />
          <BaseButton onPress={this.generatePassword} buttonText="Generate Password" />
        </View>
      );
    }
}
