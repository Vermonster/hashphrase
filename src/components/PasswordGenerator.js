import React from 'react';
import {
  Keyboard, Clipboard, View, StyleSheet, Text,
} from 'react-native';
import { withNamespaces } from 'react-i18next';
import { Button, IconButton } from 'react-native-paper';
import loplop from 'loplop';
import ConfirmationDialog from './ConfirmationDialog';
import FormInput from './FormInput';
import { colors, rowCenter } from '../styles/base';

const styles = StyleSheet.create({
  generatorContainer: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  warning: {
    flexDirection: 'row',
    width: '80%',
    justifyContent: 'space-between',
    marginTop: '-10%',
  },
  icon: {
    marginHorizontal: 0,
  },
});

class PasswordGenerator extends React.Component {
  state = {
    label: '',
    password: '',
    confirmPassword: '',
    generatedPassword: '',
    disabled: true,
    modalVisibility: false,
    inputError: false,
  };

  componentDidUpdate(prevProps) {
    const { isNewPassword } = this.props;
    if (isNewPassword !== prevProps.isNewPassword) {
      this.handleSubmitButtonState();
    }
  }

  showModal = () => this.setState({ modalVisibility: true });

  closeModal = () => this.setState({ modalVisibility: false });

  handleSubmit = (e) => {
    e.preventDefault();
    const { label, password, confirmPassword } = this.state;
    if ((password && confirmPassword) && (password !== confirmPassword)) {
      this.setState({ inputError: true });
    } else {
      this.setState({ inputError: false }, this.generatePassword(label, password));
    }
  };

  addToClipboard = (password) => {
    Clipboard.setString(password);
  };

  handleChange = name => (e) => {
    const inputValue = e.nativeEvent.text;
    this.setState({ [name]: inputValue, inputError: false }, this.handleSubmitButtonState);
  }

  handleSubmitButtonState = () => {
    this.setState({ disabled: true });
    const { label, password, confirmPassword } = this.state;
    const { isNewPassword } = this.props;
    return ((isNewPassword && label && password && confirmPassword)
    || (!isNewPassword && label && password)) ? this.setState({ disabled: false }) : null;
  }

  generatePassword(type, password) {
    this.setState({ generatedPassword: loplop(type, password) }, () => {
      const { generatedPassword } = this.state;
      this.addToClipboard(generatedPassword);
    });
    Keyboard.dismiss();
    this.showModal();
  }

  confirmEntry(inputType) {
    const { label, password, confirmPassword } = this.state;
    let status = null;
    if (inputType === 'label') {
      status = (label.length > 0);
    } else if (inputType === 'password') {
      status = (password.length > 0);
    } else if (inputType === 'confirmPassword') {
      status = (password === confirmPassword);
    }
    this.setState((prevProps) => {
      
    });
    return status;
  }

  render() {
    const {
      generatedPassword, disabled, modalVisibility, inputError, label, password, confirmPassword,
    } = this.state;
    const { isNewPassword, showSnackbar, t } = this.props;

    return (
      <View style={styles.generatorContainer}>
        <FormInput
          value={label}
          componentType="label"
          placeholderText={t('label')}
          prompt={t('passwordLabel')}
          label={t('label')}
          handleChange={this.handleChange}
          onBlur={this.confirmEntry}
        />
        <FormInput
          value={password}
          componentType="password"
          placeholderText={t('password')}
          prompt={t('masterPassword')}
          label={t('password')}
          handleChange={this.handleChange}
        />
        { isNewPassword && (
          <>
            <FormInput
              value={confirmPassword}
              componentType="confirmPassword"
              placeholderText={t('confirmPassword')}
              prompt={t('retypePassword')}
              label={t('confirmPassword')}
              error={inputError}
              handleChange={this.handleChange}
              style={styles.textInput}
            />
            { inputError && (
              <View style={[rowCenter, styles.warning]}>
                <Text style={{ color: colors.warning }}>{t('warning')}</Text>
                <IconButton icon="warning" color={colors.warning} style={styles.icon} />
              </View>
            )
            }
          </>
        )
      }
        <Button
          testID="submit-create-password"
          onPress={this.handleSubmit}
          accessibilityLabel="CREATE ACCOUNT PASSWORD BUTTON"
          disabled={disabled}
          mode="contained"
          dark
        >
          {t('submitMessage')}
        </Button>
        <ConfirmationDialog
          visible={modalVisibility}
          generatedPassword={generatedPassword}
          closeModal={this.closeModal}
          showSnackbar={showSnackbar}
        />
      </View>
    );
  }
}

export default withNamespaces('passwordGenerator')(PasswordGenerator);

export { PasswordGenerator as TestPasswordGenerator };
