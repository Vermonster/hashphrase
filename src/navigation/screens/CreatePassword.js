import React, { Component } from 'react';
import { Switch, Snackbar, IconButton } from 'react-native-paper';
import {
  View,
  Text,
  StyleSheet,
  TouchableWithoutFeedback,
  Keyboard,
  KeyboardAvoidingView,
  ScrollView,
  StatusBar,
  Dimensions,
} from 'react-native';
import { SafeAreaView } from 'react-navigation';
import { withNamespaces } from 'react-i18next';
import LogoTitle from '../../components/LogoTitle';
import PasswordGenerator from '../../components/PasswordGenerator';
import { colors, rowCenter, fontSize } from '../../styles/base';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.background,
  },
  innerContainer: {
    flex: 1,
    justifyContent: 'flex-end',
  },
  newPass: {
    ...rowCenter,
  },
  newPassLabel: {
    fontSize: fontSize.lg,
    marginRight: '2%',
  },
  snackbar: {
    flex: 1,
    justifyContent: 'flex-start',
    backgroundColor: colors.secondary,
  },
  flexLayout: {
    flex: 1,
  },
});

class CreateNewPassword extends Component {
  static navigationOptions = ({ navigation }) => ({
    headerTitle: <LogoTitle />,
    headerStyle: {
      elevation: 0,
      shadowOpacity: 0.8,
      borderBottomWidth: 0,
      shadowColor: colors.tipsPageBorder,
      shadowRadius: 5,
      shadowOffset: { width: 0, height: 3 },
      backgroundColor: colors.secondary,
    },
    headerRight: (
      <IconButton
        icon="help"
        size={30}
        color={colors.primary}
        onPress={() => navigation.navigate('TipsPage')}
      />
    ),
  });

  state = {
    isNewPassword: false,
    snackbarVisibility: false,
    height: Dimensions.get('window').height,
  };

  onLayout = () => {
    const { height } = Dimensions.get('screen');
    this.setState({ height });
  }

  handleToggleSwitch = () => this.setState(({ isNewPassword }) => ({
    isNewPassword: !isNewPassword,
  }))

  showSnackbar = () => this.setState({ snackbarVisibility: true });

  hideSnackbar = () => this.setState({ snackbarVisibility: false });

  render() {
    const { t } = this.props;
    const {
      isNewPassword,
      snackbarVisibility,
      height,
    } = this.state;

    const marginTopValue = Math.floor(height * (0.15));
    const marginBottomValue = Math.floor(height * (0.1));

    return (
      <SafeAreaView
        style={styles.container}
        onLayout={this.onLayout}
      >
        <StatusBar barStyle="light-content" />
        <KeyboardAvoidingView
          behavior="padding"
          keyboardVerticalOffset={100}
          style={styles.flexLayout}
        >
          <ScrollView style={styles.flexLayout} keyboardShouldPersistTaps="always">
            <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
              <View style={styles.innerContainer}>
                <View
                  style={[styles.newPass,
                    { marginTop: marginTopValue, marginBottom: marginBottomValue }]
                  }
                >
                  <Text style={styles.newPassLabel}>{t('newPassword')}</Text>
                  <Switch
                    value={isNewPassword}
                    onValueChange={this.handleToggleSwitch}
                    color={colors.primary}
                  />
                </View>
                <PasswordGenerator
                  isNewPassword={isNewPassword}
                  showSnackbar={this.showSnackbar}
                />
                <View style={styles.flexLayout} />
              </View>
            </TouchableWithoutFeedback>
          </ScrollView>
        </KeyboardAvoidingView>
        <Snackbar
          visible={snackbarVisibility}
          onDismiss={this.hideSnackbar}
          duration={2000}
          style={styles.snackbar}
        >
          {t('clipboardCleared')}
        </Snackbar>
      </SafeAreaView>
    );
  }
}

export default withNamespaces('createPassword')(CreateNewPassword);
export { CreateNewPassword as TestCreatePassword };
