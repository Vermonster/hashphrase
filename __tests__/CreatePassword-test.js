import React from 'react';
import { shallow } from 'enzyme';
import { View } from 'react-native';
import { Appbar } from 'react-native-paper';
import CreatePassword from '../src/navigation/screens/CreatePassword';
import PasswordGenerator from '../src/components/PasswordGenerator';

describe('/CreatePassword />', () => {
  it('should render the same snapshot', () => {
    const wrapper = shallow(<CreatePassword />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should contain all of the required components', () => {
    const wrapper = shallow(<CreatePassword />);
    const requiredComponents = [
      Appbar.Header,
      Appbar.Content,
      View,
      PasswordGenerator,
    ];

    requiredComponents.forEach((component) => {
      expect(wrapper.find(component)).toHaveLength(1);
    });
  });

  it('should toggle state when showNotification is called', () => {
    const wrapper = shallow(<CreatePassword />);
    const instance = wrapper.instance();
    expect(wrapper.state().showCopy).toEqual(false);
    instance.showNotification();
    expect(wrapper.state().showCopy).toEqual(true);
  });
});
