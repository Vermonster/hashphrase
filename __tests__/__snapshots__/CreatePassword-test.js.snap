// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreatePassword /> should render the same snapshot 1`] = `
<withOrientation
  style={
    Object {
      "backgroundColor": "#F5F5F5",
      "flex": 1,
    }
  }
>
  <StatusBar
    animated={false}
    barStyle="light-content"
    showHideTransition="fade"
  />
  <KeyboardAvoidingView
    behavior="padding"
    enabled={true}
    keyboardVerticalOffset={100}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ScrollViewMock
      keyboardShouldPersistTaps="always"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <TouchableWithoutFeedback
        onPress={[Function]}
      >
        <ForwardRef(View)
          style={
            Object {
              "flex": 1,
              "justifyContent": "flex-end",
              "marginTop": 50,
            }
          }
        >
          <LoadNamespace(PasswordGenerator) />
          <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
            style={
              Object {
                "alignItems": "center",
                "marginTop": 30,
              }
            }
          >
            <withTheme(Text)
              style={
                Object {
                  "textDecorationLine": "underline",
                }
              }
            >
              clearClipboardButton
            </withTheme(Text)>
          </TouchableOpacity>
        </ForwardRef(View)>
      </TouchableWithoutFeedback>
    </ScrollViewMock>
  </KeyboardAvoidingView>
  <withTheme(Snackbar)
    duration={2000}
    onDismiss={[Function]}
    style={
      Object {
        "backgroundColor": "#212A59",
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
    visible={false}
  >
    clipboardCleared
  </withTheme(Snackbar)>
</withOrientation>
`;
